numpy/core/tests/test_dtype.py

```diff
@@ -446,6 +447,149 @@ def test_alignment(self):
+        # Create dummy objects, defining `one` and `zero` seemed to
+        # stabilize their reference count.
+        # The test assumes that 0, 1, and None are singletons.
+        singleton = object()
+        one = 1
+        zero = 0
+        for dt, pat, count in self.iter_struct_object_dtypes(singleton):
+            for func, obj in zip([np.empty, np.ones, np.zeros],
+                                 [None, one, zero]):
+                if obj is None:
+                    # None seems flaky due to python behaviour, None's are
+                    # probably simply used in too many occasions.
+                    continue
+                gc.collect()
+                before = sys.getrefcount(obj)
+                arr = func(3, dt)
+
+                now = sys.getrefcount(obj)
+                assert now - before == count * 3
+                del arr
+                now = sys.getrefcount(obj)
+                assert now == before
+
+    def test_structured_object_item_setting(self):
+        """Structured object reference counting for simple item setting"""
+        singleton = object()
+        one = 1
+
+        for dt, pat, count in self.iter_struct_object_dtypes(singleton):
+            gc.collect()
+            before = sys.getrefcount(singleton)
+            arr = np.array([pat] * 3, dt)
+            assert sys.getrefcount(singleton) - before == count * 3
+            # Fill with `1` and check that it was replaced correctly:
+            before2 = sys.getrefcount(one)
+            arr[...] = one
+            after2 = sys.getrefcount(one)
+            assert after2 - before2 == count * 3
+            del arr
+            gc.collect()
+            assert sys.getrefcount(one) == before2
+            assert sys.getrefcount(singleton) == before
+
+    def test_structured_object_indexing(self):
+        """Structured object reference counting for advanced indexing."""
+        zero = 0
+        one = 1
+
+        shapes = [(3,), (3, 2), (3, 2)]
+        indices = [([0, 2],), ([0, 2], slice(None)), ([0, 2], [1])]
+        items_changed = [2, 4, 2]
+
+        for dt, _, count in self.iter_struct_object_dtypes():
+            for shape, index, changed in zip(shapes, indices, items_changed):
+                arr = np.zeros(shape, dt)
+
+                gc.collect()
+                before_zero = sys.getrefcount(zero)
+                before_one = sys.getrefcount(one)
+                # Test item getting:
+                part = arr[index]
+                after_zero = sys.getrefcount(zero)
+                assert after_zero - before_zero == count * changed
+                del part
+                # Test item setting:
+                arr[index] = one
+                gc.collect()
+                after_zero = sys.getrefcount(zero)
+                after_one = sys.getrefcount(one)
+                assert before_zero - after_zero == count * changed
+                assert after_one - before_one == count * changed
```

Comment:
_might_ be neat to introduce a helper object for this
```
class reference_counter:
    def __init__(self, obj):
        self._obj = obj
        self._initial_count = sys.getrefcount(obj)

    @property
    def change(self):
        return sys.getrefcount(self._obj) - self._initial_count


zero_counter = reference_counter(zero)
...
assert reference_counter.change == count * changed
```

Reply:
Yeah, possibly. Anyway, did most (not all) of the parameterization, didn't use it much yet. If you got more, can fix it up a bit more, but probably won't continue otherwise.

Maybe should add a test for the non-contiguous subarray thing at least. But it is getting out of scope to start fixing that as well and getting late today at least.